src/services/item/plugins/embeddedLink/controller.ts

```diff
@@ -0,0 +1,138 @@
+      if (!link) {
+        throw new LinkQueryParameterIsRequired();
+      }
+
+      const url = ensureProtocol(link);
+      const metadata = await embeddedLinkService.getLinkMetadata(url);
+      const isEmbeddingAllowed = await embeddedLinkService.checkEmbeddingAllowed(url, log);
+
+      return {
+        ...metadata,
+        isEmbeddingAllowed,
+      };
+    },
+  );
+
+  fastify.post(
+    '/',
+    {
+      schema: createLink,
+      preHandler: [isAuthenticated, matchOne(validatedMemberAccountRole)],
+    },
+    async (request, reply) => {
+      const {
+        user,
+        query: { parentId, previousItemId },
+        body: data,
+      } = request;
+      const member = asDefined(user?.account);
+      assertIsMember(member);
+
+      const item = await db.transaction(async (manager) => {
+        const repositories = buildRepositories(manager);
+        const item = await embeddedLinkService.postWithOptions(member, repositories, {
+          ...data,
+          previousItemId,
+          parentId,
+        });
+        return item;
+      });
+
+      reply.send(item);
+
+      // background operations
+      await actionItemService.postPostAction(request, buildRepositories(), item);
+      await db.transaction(async (manager) => {
+        const repositories = buildRepositories(manager);
+        await embeddedLinkService.rescaleOrderForParent(member, repositories, item);
+      });
+    },
+  );
+
+  fastify.patch(
+    '/:id',
+    {
+      schema: updateLink,
+      preHandler: [isAuthenticated, matchOne(validatedMemberAccountRole)],
+    },
+    async (request) => {
+      const {
+        user,
+        params: { id },
+        body,
+      } = request;
+      const member = asDefined(user?.account);
+      assertIsMember(member);
+      return await db.transaction(async (manager) => {
+        const repositories = buildRepositories(manager);
+        const item = await embeddedLinkService.patchWithOptions(member, repositories, id, body);
+        await actionItemService.postPatchAction(request, repositories, item);
+        return item;
+      });
+    },
+  );
+
+  // necessary for legacy POST /items to work with links
+  // register pre create handler to pre fetch link metadata
+  const hook = async (_actor: Actor, _repos: Repositories, { item }: { item: Partial<Item> }) => {
+    // if the extra is undefined or it does not contain the embedded link extra key, exit
+    if (!item.extra || !(ItemType.LINK in item.extra)) {
+      return;
+    }
+    const { embeddedLink } = item.extra;
+
+    const { url } = embeddedLink;
+    const { description, html, thumbnails, icons } = await embeddedLinkService.getLinkMetadata(url);
+
+    if (description) {
+      embeddedLink.description = description;
+    }
+    if (html) {
+      embeddedLink.html = html;
+    }
+
+    embeddedLink.thumbnails = thumbnails;
+    embeddedLink.icons = icons;
+
+    // default settings
+    item.settings = {
```

Comment:
Comment: I notice that we are mutating the inputs... this gives me uneasy feelings.